codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -252,7 +252,7 @@ public int onDataRead(final ChannelHandlerContext ctx, int streamId, ByteBuf dat
             Http2LocalFlowController flowController = flowController();
             try {
                 // If we should apply flow control, do so now.
-                if (shouldApplyFlowControl) {
+                if (shouldApplyFlowControl && bytesToReturn > 0) {
```

Comment:
I think we should fix the [assertion statement](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java#L312) instead to accept 0 (i.e. `assert dataLength >= 0`).  Not because 0 is expected to impact flow control but the `endOfStream` still needs to be processed.  If more issues fall out from this we should investigate those too.